libnss_determined/test/unit_test.c

```diff
@@ -0,0 +1,425 @@
+    strncpy(line, "sudo:x:1:\n", sizeof(line));
+    status = parse_group_line(line, strlen(line), sizeof(line), &result);
+    EXPECT_STATUS(status, PARSE_OK);
+    EXPECT_STRING(result.gr_name, "sudo");
+    EXPECT_STRING(result.gr_passwd, "x");
+    EXPECT_NULL_STRING(result.gr_mem[0]);
+
+    // No commas in membership list.
+    strncpy(line, "sudo:x:1:a\n", sizeof(line));
+    status = parse_group_line(line, strlen(line), sizeof(line), &result);
+    EXPECT_STATUS(status, PARSE_OK);
+    EXPECT_STRING(result.gr_name, "sudo");
+    EXPECT_STRING(result.gr_passwd, "x");
+    EXPECT_STRING(result.gr_mem[0], "a");
+    EXPECT_NULL_STRING(result.gr_mem[1]);
+
+    // No segfault on empty strings.
+    strncpy(line, "", sizeof(line));
+    status = parse_group_line(line, strlen(line), sizeof(line), &result);
+    EXPECT_STATUS(status, PARSE_INVALID);
+
+    // Buffer too short.
+    strncpy(line, "sudo:x:1:\n", sizeof(line));
+    status = parse_group_line(line, strlen(line),
+                              (strlen(line) + 1) + sizeof(*result.gr_mem) - 1,
+                              &result);
+    EXPECT_STATUS(status, PARSE_SHORT_BUFFER);
+
+    strncpy(line, "sudo:x:1:a\n", sizeof(line));
+    status = parse_group_line(
+        line, strlen(line), (strlen(line) + 1) + 2 * sizeof(*result.gr_mem) - 1,
+        &result);
+    EXPECT_STATUS(status, PARSE_SHORT_BUFFER);
+
+    // Buffer just long enough.
+    strncpy(line, "sudo:x:1:a\n", sizeof(line));
+    status = parse_group_line(
+        line, strlen(line), (strlen(line) + 1) + 2 * sizeof(*result.gr_mem) - 1,
+        &result);
+    EXPECT_STATUS(status, PARSE_SHORT_BUFFER);
+
+    // Buffer just long enough.
+    strncpy(line, "sudo:x:1:a\n", sizeof(line));
+    status = parse_group_line(line, strlen(line),
+                              (strlen(line) + 1) + 2 * sizeof(*result.gr_mem),
+                              &result);
+    EXPECT_STATUS(status, PARSE_OK);
+    EXPECT_STRING(result.gr_name, "sudo");
+    EXPECT_STRING(result.gr_passwd, "x");
+    EXPECT_STRING(result.gr_mem[0], "a");
+    EXPECT_NULL_STRING(result.gr_mem[1]);
+
+    // Ignore empty entries in membership list.
+    strncpy(line, "sudo:x:1:a,,\n", sizeof(line));
+    status = parse_group_line(line, strlen(line),
+                              (strlen(line) + 1) + 2 * sizeof(*result.gr_mem),
+                              &result);
+    EXPECT_STATUS(status, PARSE_OK);
+    EXPECT_STRING(result.gr_name, "sudo");
+    EXPECT_STRING(result.gr_passwd, "x");
+    EXPECT_STRING(result.gr_mem[0], "a");
+    EXPECT_NULL_STRING(result.gr_mem[1]);
+
+    return 0;
+}
+
+#define RUN_TEST(name, namestr)                        \
+    do {                                               \
+        if (name()) {                                  \
+            fprintf(stderr, "%s() failed\n", namestr); \
+            failed = true;                             \
+        }                                              \
+    } while (0)
```

Comment:
non-blocking: Can save the extra argument using stringification:
```suggestion
#define RUN_TEST(name)                            \
    do {                                          \
        if (name()) {                             \
            fprintf(stderr, #name "() failed\n"); \
            failed = true;                        \
        }                                         \
    } while (0)
```
And then just
```c
    RUN_TEST(test_get_field);
    // etc.
```

Reply:
Heh, apparently GCC now refers to this operation as "[stringizing](https://gcc.gnu.org/onlinedocs/gcc-7.5.0/cpp/Stringizing.html)" rather than "[stringification](https://gcc.gnu.org/onlinedocs/gcc-6.5.0/cpp/Stringification.html)". :man_shrugging:

Reply:
done.